File: Aula 1/JS/6_polimorfismo_sobrecarga.js
// NOTA: JS não permite duas funções com mesmo nome na mesma classe, 
// então implementação do polimorfismo de sobrecarga não é tão simples
// como em C++ por exemplo.

// Classe
class Calculadora {   
    // Metodo soma tres numeros
    soma (a, b, c) {
        if (c === undefined)
            return a + b;
        else 
            return a + b + c;
    }
}

// Objetos
let Calc = new Calculadora();

console.log(Calc.soma(1, 2));
console.log(Calc.soma(1, 2, 3));

File: Aula 1/C#/6_polimorfismo_sobrecarga.cs
namespace Course
{
    // NOTA: Diferente do JavaScript, o C# permite múltiplos métodos com o
    // mesmo nome na mesma classe, desde que com assinaturas diferentes

    // Classe
    internal class Calculadora
    {
        // Método soma dois números
        public double Soma(double a, double b)
        {
            return a + b;
        }

        // Sobrecarga: mesmo nome, assinatura diferente (três parâmetros)
        public double Soma(double a, double b, double c)
        {
            return a + b + c;
        }
    }

    internal class Program
    {
        static void Main(string[] args)
        {
            // Objeto
            Calculadora calc = new Calculadora();

            // O compilador escolhe a versão correta pela quantidade de argumentos
            Console.WriteLine(calc.Soma(1, 2));
            Console.WriteLine(calc.Soma(1, 2, 3));
        }
    }
}
